extensions/csv/src/main/java/io/deephaven/csv/tokenization/Tokenizer.java

```diff
@@ -0,0 +1,530 @@
+package io.deephaven.csv.tokenization;
+
+import gnu.trove.map.hash.TIntObjectHashMap;
+import io.deephaven.csv.tokenization.external.DoubleParser;
+import io.deephaven.csv.containers.ByteSlice;
+import io.deephaven.time.TimeZone;
+import org.joda.time.DateTime;
+import org.joda.time.DateTimeZone;
+
+public class Tokenizer {
+    /**
+     * Temporary variable owned by tryParseDBDateTime.
+     */
+    private final long[] dbDateTimeTemp0 = new long[1];
+    /**
+     * Temporary variable owned by tryParseDBDateTime.
+     */
+    private final long[] dbDateTimeTemp1 = new long[1];;
+    /**
+     * Temporary variable owned by tryParseDBDateTime.
+     */
+    private final long[] dbDateTimeTemp2 = new long[1];;
+    /**
+     * Temporary variable owned by tryParseDBDateTime.
+     */
+    private final DateTimeZone[] dbDateTimeTempZone = new DateTimeZone[1];
+
+    private final TimeZoneParser timeZoneParser = new TimeZoneParser();
+
+    /**
+     * ByteSlice is *NOT* modified, regardless of sucess or failure.
+     */
+    public boolean tryParseBoolean(ByteSlice bs, boolean[] result) {
+        final byte[] d = bs.data();
+        final int o = bs.begin();
+        final int bSize = bs.end() - o;
+        if (bSize == 4) {
+            result[0] = true; // Optimistic
+            // sad
+            return (d[o] == 't' || d[o] == 'T') &&
+                    (d[o + 1] == 'r' || d[o + 1] == 'R') &&
+                    (d[o + 2] == 'u' || d[o + 2] == 'U') &&
+                    (d[o + 3] == 'e' || d[o + 3] == 'E');
+        }
+
+        if (bSize == 5) {
+            result[0] = false; // Optimistic
+            // mega sad
+            return (d[o] == 'f' || d[o] == 'F') &&
+                    (d[o + 1] == 'a' || d[o + 1] == 'A') &&
+                    (d[o + 2] == 'l' || d[o + 2] == 'L') &&
+                    (d[o + 3] == 's' || d[o + 3] == 'S') &&
+                    (d[o + 4] == 'e' || d[o + 4] == 'E');
+        }
+
+        return false;
+    }
+
+    /**
+     * ByteSlice is *NOT* modified, regardless of sucess or failure.
+     */
+    public boolean tryParseLong(final ByteSlice bs, long[] result) {
+        final int savedBegin = bs.begin();
+        Mutating.trim(bs);
+        final boolean success = Mutating.tryParseLong(bs, result);
+        bs.setBegin(savedBegin);
+        return success;
+    }
+
+    /**
+     * ByteSlice is *NOT* modified, regardless of sucess or failure.
+     */
+    public boolean tryParseDouble(ByteSlice bs, double[] result) {
+        return DoubleParser.tryParseDouble(bs, result);
+    }
+
+    public boolean hasFewerThan8SignificantFigures(ByteSlice bs) {
+        final byte[] d = bs.data();
+        int current = bs.begin();
+        final int end = bs.end();
+
+        // Find first digit
```

Comment:
I think I would agree, except that the variables are used for *both* while loops, and so I thought it would be better to have them offset from the whiles. Let me know if you agree/disagree.

Reply:
Agreed.